internal/datanode/timetick_sender.go

```diff
@@ -0,0 +1,173 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus-proto/go-api/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/msgpb"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/types"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/commonpbutil"
+	"github.com/milvus-io/milvus/pkg/util/retry"
+)
+
+// timeTickSender is to merge channel states updated by flow graph node and send to datacoord periodically
+// timeTickSender hold a SegmentStats time sequence cache for each channel,
+// after send succeeds will clean the cache earlier than the sended timestamp
+type timeTickSender struct {
+	nodeID    int64
+	dataCoord types.DataCoord
+
+	mu                  sync.Mutex
+	channelStatesCaches map[string]*segmentStatesSequence // string -> *segmentStatesSequence
+}
+
+// data struct only used in timeTickSender
+type segmentStatesSequence struct {
+	data map[uint64][]*commonpb.SegmentStats // ts -> segmentStats
+}
+
+func newTimeTickManager(dataCoord types.DataCoord, nodeID int64) *timeTickSender {
+	return &timeTickSender{
+		nodeID:              nodeID,
+		dataCoord:           dataCoord,
+		channelStatesCaches: make(map[string]*segmentStatesSequence, 0),
+	}
+}
+
+func (m *timeTickSender) start(ctx context.Context) {
+	ticker := time.NewTicker(Params.DataNodeCfg.DataNodeTimeTickInterval.GetAsDuration(time.Millisecond))
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("timeTickSender context done")
+			return
+		case t := <-ticker.C:
+			m.sendReport(ctx, uint64(t.Unix()))
+		}
+	}
+}
+
+func (m *timeTickSender) update(channelName string, timestamp uint64, segmentStats []*commonpb.SegmentStats) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	channelStates, ok := m.channelStatesCaches[channelName]
+	if !ok {
+		channelStates = &segmentStatesSequence{
+			data: make(map[uint64][]*commonpb.SegmentStats, 0),
+		}
+	}
+	channelStates.data[timestamp] = segmentStats
+	m.channelStatesCaches[channelName] = channelStates
+}
+
+func (m *timeTickSender) mergeDatanodeTtMsg() ([]*msgpb.DataNodeTtMsg, map[string]uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var msgs []*msgpb.DataNodeTtMsg
+	sendedLastTss := make(map[string]uint64, 0)
+
+	for channelName, channelSegmentStates := range m.channelStatesCaches {
+		var lastTs uint64
+		segmentStats := make([]*commonpb.SegmentStats, 0)
+		for ts, segmentStates := range channelSegmentStates.data {
```

Comment:
should we only send an maintained the lastest segmentStats for each segment? 
While merge I think it's a good chance to overwrite old segment states with new Segment states